### ISSUE TITLE
FCN的操作步骤

### ISSUE THREAD

**Issue by 413180794** [413180794/-#2]
训练fcn：
1、根目录/home/xjy/fcn.berkeleyvision.org/下创建fcn32s文件夹，准备文件
fcn32s里面有：

Infer.py  外层                                                            score.py   外层
Solve.py   已有
Solver.prototxt  已有
sUrgery.py    外层
Train.prototxt   已有
Val.prototxt   已有
voc_helper.py外层
voc_layers.py外城
Deploy.prototxt 已有
Net.py 已有

2.在fcn32s下，创建文件夹11 snapshot：mkdir snapshot

3.在/home/xjy/fcn.berkeleyvision.org/data/下，
（1）创建存放数据集的文件夹： mkdir taibiao_datasets_45_classes
（2）在taibiao_datasets_45_classes中创建3个文件夹：mkdir ImageSets JPEGImages SegmentationClass
（4）将图像存入JPEGImages，将标签存入SegmentationClass；（待做）
（5） 在ImageSets下，创建文件夹：mkdir Segmentation
（6）在 Segmentation中存放文本：train.txt val.txt，根据图像生成train.txt、val.txt（待做，即使将图像的路径保存在文本文档）
 
4.修改train.prototxt：
（1）修改数据层：
  layer {                                                                                                                                                                                    
  2   name: "data"
  3   type: "Python"
  4   top: "data"
  5   top: "label"
  6   python_param {
  7     module: "voc_layers"
  8     layer: "VOCSegDataLayer"
  9     param_str: "{\'voc_dir\': \'/home/windowsDisk/zhfanData/caffe/fcn.berkeleyvision.org/data/taibiao_8_classes\', \'seed\': 1337, \'split\': \'train\', \'mean\': (104.00699, 116.66877, 1    22.67892)}"
 10   }
 11 }修改第一层
                                                                                                                           11 }

（2）修改输出类别数：
将所有的num_output: 21替换为num_output: 46
（45类+1个背景类）（2个地方）

修改val.prototxt：与train.prototxt一样的修改；
layer {                                                                                                                                                                                    
  2   name: "data"
  3   type: "Python"
  4   top: "data"
  5   top: "label"
  6   python_param {
  7     module: "voc_layers"
  8     layer: "VOCSegDataLayer"
  9     param_str: "{\'voc_dir\': \'/home/windowsDisk/zhfanData/caffe/fcn.berkeleyvision.org/data/taibiao_8_classes\', \'seed\': 1337, \'split\': \'val\', \'mean\': (104.00699, 116.66877, 122    .67892)}"
 10   }
 11 }
修改第一层 
（2）修改输出类别数：
将所有的num_output: 21替换为num_output: 46
（45类+1个背景类）（2个地方）

6.修改solver.prototxt：根据训练样本数量修改训练次数；（待改）

  1 train_net: "train.prototxt"                                                                                                                                                                
  2 test_net: "val.prototxt"
  3 test_iter: 736
  4 # make test net, but don't invoke it from the solver itself
  5 test_interval: 999999999
  6 display: 20
  7 average_loss: 20
  8 lr_policy: "fixed"
  9 # lr for unnormalized softmax
 10 base_lr: 1e-10
 11 # high momentum
 12 momentum: 0.99
 13 # no gradient accumulation
 14 iter_size: 1
 15 max_iter: 7000 #100000（一次训练7000张）
 16 weight_decay: 0.0005
 17 snapshot: 1000（每训练1000次保存一次模型）
 18 snapshot_prefix: "snapshot/train"
 19 test_initialization: false
~                                                                                                                                                                                              
~                                      
7.修改solve.py：
  import surgery, score
  3 
  4 import numpy as np
  5 import os
  6 import sys
  7 sys.path.append('/home/windowsDisk/zhfanData/caffe/python')
  8 import caffe
  9 
 10 #weights = '../ilsvrc-nets/vgg16-fcn.caffemodel'（训练好的 32s）
 11 vgg_weights = '../ilsvrc-nets/vgg16-fcn.caffemodel'
 12 vgg_proto = '../ilsvrc-nets/VGG_ILSVRC_16_layers_deploy.prototxt'（32s）
 13 
 14 
 15 
 16 # init
 17 caffe.set_device(0)
 18 caffe.set_mode_gpu()
 19 
 20 solver = caffe.SGDSolver('solver.prototxt')
 21 #solver.net.copy_from(weights)（下一个）
 22 vgg_net = caffe.Net(vgg_proto,vgg_weights,caffe.TRAIN)（32）
 23 surgery.transplant(solver.net,vgg_net)（32）
 24 del vgg_net（32）
 25 
 26 
 27 
 28 # surgeries
 29 interp_layers = [k for k in solver.net.params.keys() if 'up' in k]
 30 surgery.interp(solver.net, interp_layers)
# scoring
 33 val = np.loadtxt('../data/taibiao_8_classes/ImageSets/Segmentation/val.txt', dtype=str)
 34 
 35 for _ in range(7):
 36     solver.step(1000)
 37     score.seg_tests(solver, False, val, layer='score')  

----------------------------------------------------------------------------------------------------------------------------
FCN测试
1.根据train.prototxt，生成一个deploy.prototxt：
deploy.prototxt用于测试的文件：
（1）数据层修改：
  1 layer {                                                                                                                         
  2   name: "input"
  3   type: "Input"
  4   top: "data"
  5   input_param {
  6     shape { dim: 1 dim: 3 dim: 500 dim: 500 }  #根据测试数据大小修改（待改）
  7   }
  8 }
（2）去掉train.prototxt 最后的分类层；

2.修改infer.py：
 14 fin = open('/home/xjy/fcn.berkeleyvision.org/data/taibiao_datasets_45_classes/ImageSets/Segmentation/trainval.txt','r')
 15 image_lists = fin.read().splitlines()
测试文本：trainval.txt（待做）

18 net = caffe.Net('deploy.prototxt', 'snapshot/train_iter_100000.caffemodel', caffe.TEST) （待修改caffemodel）
Nohup python solve.py > .txt
vim
